File: leet_easy_python/leet_020/leet_020.py
class Solution(object):
    def isValid(self, s):
        stk = []
        sz = len(s)
        i = 0
        while i<sz:
            c = s[i];
            if c=='(' or c=='[' or c=='{':
                stk.append(c)
            elif len(stk) == 0: return False
            elif c==')' and stk[len(stk)-1]=='(': stk.pop()
            elif c==']' and stk[len(stk)-1]=='[': stk.pop()
            elif c=='}' and stk[len(stk)-1]=='{': stk.pop()
            else: return False
            i+=1
        if len(stk) != 0: return False
        return True
            
        """
        :type s: str
        :rtype: bool
        """
        

File: leet_easy_javascript/leet_020/leet_020.js
var Stack = function(){
    this.Stk = [];
    this.push = function(item){ this.Stk.push(item); };
    this.pop = function(){ return this.Stk.pop(); };
    this.peek = function(){ return this.Stk[this.Stk.length - 1]; };
    this.empty = function(){ return this.Stk.length === 0; };
};

var validate = {};
validate["("] = "push";
validate["["] = "push";
validate["{"] = "push";
validate[")"] = "(";
validate["]"] = "[";
validate["}"] = "{";


/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    var stk = new Stack(), tmp;
    for(var i=0; i<s.length; i++){
        tmp = validate[s[i]];
        if(tmp === undefined) return false;
        else if(tmp === "push") stk.push(s[i]);
        else if(stk.peek() !== tmp) return false;
        else stk.pop();
    }
    return stk.empty();
};